Warn about not implementing complex calculations in DataGrid's calculateCellValue (#3)
DevExpress/devextreme-documentation: api-reference/_hidden/GridBaseColumn/calculateCellValue.md

---

#include uiwidgets-ref-functioncontext with { 
    value: "column's configuration"
}

> [note] This function is called multiple times for every record: when the widget renders it, applies sorting or filtering, and computes summaries. We recommend that you keep calculations inside this function as simple as possible so as not to lower the widget's performance.

Reviewer: ```suggestion
[note] This function is called multiple times for every record: when the widget renders it, when it applies sorting or filtering, and when it computes summaries. It is recommend that you keep calculations inside this function as simple as possible to avoid hindering widget performance.
```

Reviewer 2: > This function is called multiple times for every record: when the widget renders it, ...

Here, "it" refers to the record, while later:

> ... when it applies sorting or filtering, and when it computes summaries.

... "it" refers to the widget.

Reviewer: can you rewrite it?

Reviewer 2: ```suggestion
[note] This function is called multiple times for every record: when the record is rendered, when sorting or filtering is applied, and when summaries are computed. It is recommend that you keep calculations inside this function as simple as possible to avoid hindering widget performance.
```

> [important] Do not implement complex calculations within this method to avoid poor widget's performance. This method is called every time a row renders or when a sorting/filtering/summary setting changes. Thus, it can be called multiple times in a short period and multiple complex calculations can slow down the widget.

Reviewer 2: We warn the reader but do not provide a way to solve the issue. What calculation is considered complex? What to do if the reader still needs this calculation to be performed?

@vconst Can't we fix this in code? Why does this function need to be called when data is sorted, filtered, etc.? Why recalculate the values when they were already calculated?

Reviewer 3: Adding cache for such cases is not simple task. And in most of user cases cache is not necessary and it may be bad because cache takes up addition memory.

If user has large calculations in caclulateCellValue, he can add caching to his own code.

So, customers should know that caclulateCellValue is called many times and they should optimize their own code by simplifing or by adding cache.

Reviewer 2: > If user has large calculations in caclulateCellValue, he can add caching to his own code.

Any good articles about caching to which we could forward the reader?

Reviewer 3: I do not have any specific links for this theme
